phac-nml/irida-next
# [STRY0011998] Linelist Export Service and Job

## What does this PR do and why?
This PR adds linelist (metadata) exports to the data exports `create_service` and `create_job`.

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
1. Reseed your database if needed with: `bin/rails db:drop/create/migrate/seed`
2. Test CSV and `namespace_type: 'Project'`
  - In `rails c` assign the following:
  
```
user = User.second
params_1 = {'export_type' => 'linelist', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Streptococcus pyogenes M92/Outbreak 2023 Sample 1').id, Sample.find_by(name: 'Streptococcus pyogenes M92/Outbreak 2023 Sample 2').id, Sample.find_by(name: 'Streptococcus pyogenes M92/Outbreak 2023 Sample 3').id, Sample.find_by(name: 'Streptococcus pyogenes M92/Outbreak 2023 Sample 4').id], 'format' => 'csv', 'namespace_id' => Sample.find_by(name: 'Streptococcus pyogenes M92/Outbreak 2023 Sample 2').project.namespace.id, 'metadata_fields' => ['age', 'food', 'onset']}}
```

  - Call the service:
  
```
DataExports::CreateService.new(user, params_1).execute
```

  - After 30s, download the file (if the URL only contains `localhost`, replace it with `localhost:3000`):

```
data_export = user.data_exports.last
Rails.application.routes.url_helpers.rails_blob_url(data_export.file)
```

  - Open the CSV and ensure the expected samples and their respective IDs match, as well as the metadata values.

3. Test .xlsx, `namespace_type: 'Group'`, and varying metadata fields
  - On the website, navigate to any number of samples that `user1` can export, that belong to the same group but different projects. I will use `Streptococcus pyogenes M92/Outbreak 2023 Sample 10`. `Streptococcus pyogenes M92/Outbreak 2022 Sample 10`, and `Streptococcus pneumoniae/Outbreak 2022 Sample 10`.
  - Assign DIFFERENT metadata fields to each one. I will use `test1`, `test2`, and `test3`. Value does not matter
  - In `rails c` assign the following (ensure you use the correct IDs and metadata fields if you used a different set): 

```
user = User.second
params_1 = {'export_type' => 'linelist', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Streptococcus pyogenes M92/Outbreak 2023 Sample 10').id, Sample.find_by(name: 'Streptococcus pyogenes M92/Outbreak 2022 Sample 10').id, Sample.find_by(name: 'Streptococcus pneumoniae/Outbreak 2022 Sample 10').id], 'format' => 'xlsx', 'namespace_id' => Namespace.find_by(name: "Streptococcus").id, 'metadata_fields' => ['age', 'food', 'onset', 'test1', 'test2', 'test3']}}
```

  - Call the service:

```
DataExports::CreateService.new(user, params_1).execute
```

  - After 30s, download the file (if the URL only contains `localhost`, replace it with `localhost:3000`):

```
data_export = user.data_exports.last
Rails.application.routes.url_helpers.rails_blob_url(data_export.file)
```

  - Open the spreadsheet and ensure the following:
    - Sample names, IDs, and metadata respectively match.
    - Project paths are present and accurate
    - Metadata fields that samples DO NOT have are blank


4. Re-test Sample and Workflow Execution exports. The service and job code are shared between all exports and therefore required refactoring to fit `linelist` exports in. Instructions to run these exports are here: [sample](https://github.com/phac-nml/irida-next/pull/503) and [workflow](https://github.com/phac-nml/irida-next/pull/536)

5, Test new authorization with sample export as the logic for samples selected for export (used by sample and linelist exports) has been revamped
- Login as `user1` and add group `Vibrio` to group `Streptococcus` with an `access level` of `Guest`
- Logout and log in as `user4` (owner of vibrio)
- Navigate to the `Vibrio` group samples table. `Streptococcus` sample should be present in the table.
- Create an export which includes both `Vibrio` and `Streptococcus` samples.
- Ensure an error occurs as Vibrio users do not have high enough access to export `Streptococcus` samples.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
